ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OKMeans.scala

```diff
@@ -0,0 +1,138 @@
+        s" converted to representation H2O-3 understands.")
+    }
+  }
+
+  def this() = this(Identifiable.randomUID("kmeans"))
+}
+
+object H2OKMeans extends DefaultParamsReadable[H2OKMeans]
+
+
+/**
+  * Parameters for Spark's API exposing underlying H2O model.
+  */
+trait H2OKMeansParams extends H2OAlgoUnsupervisedParams[KMeansParameters] {
+
+  type H2O_SCHEMA = GLMParametersV3
+
+  protected def paramTag = reflect.classTag[KMeansParameters]
+
+  protected def schemaTag = reflect.classTag[H2O_SCHEMA]
+
+  //
+  // Param definitions
+  //
+  private val maxIterations = intParam("maxIterations", "Maximum number of KMeans iterations to find the centroids.")
+  private val standardize = booleanParam("standardize", "Standardize the numeric columns to have a mean of zero and unit variance.")
+  private val init = stringParam("init", "Initialization mode for finding the initial cluster centers.")
+  private val userPoints = nullableDoubleArrayArrayParam("userPoints", "This option allows" +
+    " you to specify array of points, where each point represents coordinates of an initial cluster center. The user-specified" +
+    " points must have the same number of columns as the training observations. The number of rows must equal" +
+    " the number of clusters.")
+  private val estimateK = booleanParam("estimateK", "If enabled, the algorithm tries to identify optimal number of clusters, up to k clusters.")
+  private val k = intParam("k", "Number of clusters to generate.")
+
+  //
+  // Default values
+  //
+  setDefault(
+    maxIterations -> 10,
+    standardize -> true,
+    init -> KMeans.Initialization.Furthest.name(),
+    userPoints -> null,
+    estimateK -> false,
+    k -> 2
+  )
+
+  //
+  // Getters
+  //
+  def getMaxIterations(): Int = $(maxIterations)
+
+  def getStandardize(): Boolean = $(standardize)
+
+  def getInit(): String = $(init)
+
+  def getUserPoints(): Array[Array[Double]] = $(userPoints)
+
+  def getEstimateK(): Boolean = $(estimateK)
+
+  def getK(): Int = $(k)
+
+
+  //
+  // Setters
+  //
+  def setMaxIterations(value: Int): this.type = set(maxIterations, value)
+
+  def setStandardize(value: Boolean): this.type = set(standardize, value)
+
+  def setInit(value: String): this.type = {
+    val validated = H2OAlgoParamsHelper.getValidatedEnumValue[KMeans.Initialization](value)
+    set(init, validated)
+  }
+
+  def setUserPoints(value: Array[Array[Double]]): this.type = set(userPoints, value)
+
+  def setEstimateK(value: Boolean): this.type = set(estimateK, value)
+
+  def setK(value: Int): this.type = set(k, value)
+
+  override def updateH2OParams(): Unit = {
+    super.updateH2OParams()
+    parameters._max_iterations = getMaxIterations()
+    parameters._standardize = getStandardize()
+    parameters._init = KMeans.Initialization.valueOf(getInit())
+    parameters._user_points = {
+      val userPoints = getUserPoints()
+      if (userPoints == null || userPoints.length == 0) {
```

Comment:
Just wondering whether an empty array express the same thing or not. If so, shouldn't we use an empty array as a default value?

Reply:
though bout it, it does not, my mistake, thanks!